subsys/power/device_pm.c

```diff
@@ -0,0 +1,301 @@
+{
+	u32_t current_state;
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto handler_out;
+	}
+
+	/* Check if current power state is same as target
+	 * state so that we can skip processing new request.
+	 */
+	if (current_state == target_state) {
+		if (pm_flags & DEVICE_PM_ASYNC) {
+			if (dev->pm->async_request == target_state) {
+				goto handler_out;
+			}
+		} else {
+			goto handler_out;
+		}
+	}
+
+	/* Schedule the worker for new async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->async_request = target_state;
+		dev->pm->async_req_status = DEVPM_ASYNC_REQ_QUEUED;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+
+		return DEVICE_PM_RET_QUEUED;
+	}
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, target_state);
+	if (ret < 0) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+						target_state, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+handler_out:
+	return ret;
+}
+
+static int device_pm_handle_get(struct device *dev, u32_t pm_flags)
+{
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage >= UINT_MAX) {
+		LOG_ERR("usage count overflow error!!\n");
+		ret = -ERANGE;
+		goto get_out;
+	}
+	dev->pm->usage++;
+
+	if (!dev->pm->enable) {
+		goto get_out;
+	}
+
+	if (dev->pm->usage > 0) {
+		ret = device_pm_state_handler(dev,
+				DEVICE_PM_ACTIVE_STATE, pm_flags);
+	}
+
+
+get_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+static int device_pm_handle_put(struct device *dev, u32_t pm_flags)
+{
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (dev->pm->usage <= 0) {
+		LOG_ERR("usage count underflow error!!\n");
+		ret = -ERANGE;
+		goto put_out;
+	}
+	dev->pm->usage--;
+
+	if (!dev->pm->enable) {
+		goto put_out;
+	}
+
+	if (dev->pm->usage < 1) {
+		ret = device_pm_state_handler(dev,
+				DEVICE_PM_SUSPEND_STATE, pm_flags);
+	}
+
+put_out:
+	k_sem_give(&dev->pm->lock);
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_handle_get(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_handle_get(dev, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_handle_put(dev, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_handle_put(dev, 0);
+}
+
+void device_pm_enable(struct device *dev)
+{
+	int ret;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->enable = true;
+
+	/* During the driver init, device can set the
```

Comment:
IMHO we have to have system-wide policy here.

Reply:
means? what do you suggest?

Reply:
IMHO all drivers should be obligated to leave the device in the active state after init. Per-driver policy will create a mess.

Reply:
The devices might be active but we need to call the pm_enable to init the device idle info. So the driver will need to call pm_enable during init or during config.

Reply:
This means, that every device driver will have to be updated.

Reply:
What about _sys_device_do_config_level() ?

Reply:
yeah, we can add in the for loop after device init is called.